public/app/features/alerting/unified/components/rules/ReorderRuleGroupModal.tsx

```diff
@@ -0,0 +1,233 @@
+import { css } from '@emotion/css';
+import cx from 'classnames';
+import { compact } from 'lodash';
+import React, { FC, useCallback, useState } from 'react';
+import {
+  DragDropContext,
+  Draggable,
+  DraggableProvided,
+  Droppable,
+  DroppableProvided,
+  DropResult,
+} from 'react-beautiful-dnd';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import { Badge, Icon, Modal, Tooltip, useStyles2 } from '@grafana/ui';
+import { dispatch } from 'app/store/store';
+import { CombinedRule, CombinedRuleGroup, CombinedRuleNamespace } from 'app/types/unified-alerting';
+
+import { updateRulesOrder } from '../../state/actions';
+import { isCloudRulesSource } from '../../utils/datasource';
+import { hashRulerRule } from '../../utils/rule-id';
+import { isAlertingRule, isRecordingRule } from '../../utils/rules';
+
+import { AlertStateTag } from './AlertStateTag';
+
+interface ModalProps {
+  namespace: CombinedRuleNamespace;
+  group: CombinedRuleGroup;
+  onClose: () => void;
+}
+
+type CombinedRuleWithUID = { uid: string } & CombinedRule;
+
+export const ReorderCloudGroupModal: FC<ModalProps> = (props) => {
+  const { group, namespace, onClose } = props;
+  const [pending, setPending] = useState<boolean>(false);
+  const [rulesList, setRulesList] = useState<CombinedRule[]>(group.rules);
+
+  const styles = useStyles2(getStyles);
+
+  const onDragEnd = useCallback(
+    (result: DropResult) => {
+      // check for no-ops so we don't update the group unless we have changes
+      if (!result.destination) {
+        return;
+      }
+
+      const sameIndex = result.destination.index === result.source.index;
+      if (sameIndex) {
+        return;
+      }
+
+      const newOrderedRules = reorder(rulesList, result.source.index, result.destination.index);
+      setRulesList(newOrderedRules); // optimistically update the new rules list
+
+      // the rulesSource is sometimes a string (for Grafana managed) or a rulesSource object
+      const rulesSourceName =
+        typeof namespace.rulesSource === 'string' ? namespace.rulesSource : namespace.rulesSource.name;
+
+      const rulerRules = compact(newOrderedRules.map((rule) => rule.rulerRule));
```

Comment:
Why `compact` is needed?

Reply:
Great question! 

According to the type definition the `rulerRule` is an optional property, so the `rulerRules` variable is typed as `(CombinedRule | undefined)[]` – this conflicts with what `updateRulesOrder` wants (which is `CombinedRule[]`).

If we use the `compact` function we can omit any rules without `rulerRule` and satisfy the type checker at the same time.

I couldn't get it to work with `[].filter().map()` either and didn't want to use a type cast either.